### PR TITLE
Fix #894 - badge emails and signals

### LINKED ISSUE #894: Badges notification mails are not sent (signals are not dispatched)
## Detailed Description
Badges notification mails are not sent because the badges signals ([`on_badge_added`](https://github.com/opendatateam/udata/blob/master/udata/core/badges/signals.py#L9) and [`on_badge_removed`](https://github.com/opendatateam/udata/blob/master/udata/core/badges/signals.py#L9) are never dispatched.

## Context
#885 and #887  highlighted the fact that badges metrics were not in use because signals are never dispatched.
Notifications mail are connected on the `on_badge_added` and so are never sent too.

## Possible Implementation
Dispatch the signals were they should be (on badges addition and removal)

## Your Environment
udata 1.0.x

### PR DESCRIPTION
**Pull request by abulte**
Fix #894

In addition to fixing email sending, the system is now more easily pluggable/extendable: it's easy to create a notification for a custom by subscribing to the `on_badge_added` and `on_badge_removed` signals.

### CODE REVIEW

File: udata/core/badges/tasks.py [hunk 1]
```diff
@@ -23,20 +23,57 @@ def call_task(badge, **kwargs):
...
+    sender
+        The object that emitted the event.
+    the_badge
+        The `Badge` object awarded.
+        NB: `badge` is a reserved param name in Blinker.
+    '''
+    if the_badge.kind == PUBLIC_SERVICE and isinstance(sender, Organization):
+        recipients = [member.user for member in sender.members]
+        subject = _(
+            'Your organization "%(name)s" has been identified as public service',
```
Comment (davidbgk) [inline udata/core/badges/tasks.py]: The subject does not contain `a public service` as in emails body. Not sure which one is correct.

File: udata/core/badges/models.py [hunk 2]
```diff
@@ -84,16 +86,20 @@ def add_badge(self, kind):
...
     def remove_badge(self, kind):
         '''Perform an atomic removal for a given badge'''
+        badge = self.get_badge(kind)
         self.update(__raw__={
             '$pull': {
                 'badges': {'kind': kind}
             }
         })
         self.reload()
+        on_badge_removed.send(self, the_badge=badge)
```
Comment (davidbgk) [inline udata/core/badges/models.py]: Is that signal in use somewhere?

File: udata/tests/api/test_organizations_api.py [hunk 3]
```diff
@@ -682,22 +684,59 @@ def test_list(self):
...
+        self.assertListEqual([m.recipients[0] for m in mails], members_emails)
+
+    def test_create_same(self):
+        data = self.factory.attributes()
+        with self.api_user():
+            with self.assert_emit(on_badge_added):
+                self.post(
+                    url_for('api.organization_badges', org=self.organization),
+                    data)
+            with self.assert_not_emit(on_badge_added):
```
Comment (davidbgk) [inline udata/tests/api/test_organizations_api.py]: Might be handy to test that no email has been sent in that case.

File: udata/core/badges/models.py (hunk 2 above)
Comment (abulte) [inline udata/core/badges/models.py]: Nope. Good for "plugability" though?

File: udata/tests/api/test_organizations_api.py (hunk 3 above)
Comment (abulte) [inline udata/tests/api/test_organizations_api.py]: 👍

File: udata/core/badges/tasks.py (hunk 1 above)
Comment (abulte) [inline udata/core/badges/tasks.py]: Probably `public service` (https://en.wikipedia.org/wiki/Public_service). Thanks.

**Comment by abulte** [comment]
@noirbizarre is this OK now? Thanks.

File: udata/tests/api/test_organizations_api.py [hunk 4]
```diff
@@ -682,22 +684,63 @@ def test_list(self):
     def test_create(self):
         data = self.factory.attributes()
         with self.api_user():
-            response = self.post(
-                url_for('api.organization_badges', org=self.organization),
-                data)
+            with self.assert_emit(on_badge_added):
```
Comment (noirbizarre) [inline udata/tests/api/test_organizations_api.py]: I think you can wrap this in a single context manager (if it pass the 80 chars constraint)

Comment (abulte) [inline udata/tests/api/test_organizations_api.py]: 👍

**Pull request merged**